kyo-core/shared/src/main/scala/kyo/layers.scala

```diff
@@ -0,0 +1,88 @@
+package kyo
+
+import Layers.internal.*
+import java.util.concurrent.ConcurrentHashMap as JConcurrentHashMap
+import kyo.core.*
+import scala.annotation.targetName
+
+sealed trait Layer[+Out, -S]:
+    self =>
+
+    infix def to[Out2, S2, In2](that: Layer[Out2, Envs[In2] & S2]): Layer[Out2, S & S2]          = To(self, that)
+    infix def and[Out2, S2](that: Layer[Out2, S2]): Layer[Out & Out2, S & S2]                    = And(self, that)
+    infix def using[Out2, S2, In2](that: Layer[Out2, Envs[In2] & S2]): Layer[Out & Out2, S & S2] = self and (self to that)
+
+    private[kyo] def doRun(memoMap: JConcurrentHashMap[Layer[?, ?], Any] = JConcurrentHashMap()): TypeMap[Out] < (S & IOs) =
+        type Expected = TypeMap[Out] < (S & IOs)
+        memoMap.get(self) match
+            case nullable if isNull(nullable) =>
+                self match
+                    case And(lhs, rhs) =>
+                        {
+                            for
+                                leftResult  <- lhs.doRun(memoMap)
+                                rightResult <- rhs.doRun(memoMap)
+                            yield leftResult.union(rightResult)
+                        }.asInstanceOf[Expected]
+
+                    case To(lhs, rhs) =>
+                        {
+                            for
+                                leftResult  <- lhs.doRun(memoMap)
+                                rightResult <- Envs.runTypeMap(leftResult)(rhs.doRun(memoMap))(using summon, Envs.bypass)
+                            yield rightResult
+                        }.asInstanceOf[Expected]
+
+                    case FromKyo(kyo) =>
+                        kyo().map { result => IOs(memoMap.compute(self, { (_, _) => result })).map(_ => result) }.asInstanceOf[Expected]
```

Comment:
Minor: it's possible to avoid the `map` transformation:

```scala
IOs {
    memoMap.compute(self, { (_, _) => result })
    result
}
```

By the way, couldn't we use `putIfAbsent`?